adapters/adapterstest/test_json.go

```diff
@@ -331,3 +324,110 @@ func diffJson(t *testing.T, description string, actual []byte, expected []byte)
 		}
 	}
 }
+
+// testMakeBidsImpl asserts the results of the bidder MakeRequests implementation against the expected JSON-defined results
+// and makes sure no data races occur
+func testMakeRequestsImpl(t *testing.T, filename string, spec *testSpec, bidder adapters.Bidder, reqInfo *adapters.ExtraRequestInfo) []*adapters.RequestData {
+	t.Helper()
+
+	deepBidReqCopy, shallowBidReqCopy := getDataRaceTestCopies(&spec.BidRequest)
+
+	// Run MakeRequests
+	actualReqs, errs := bidder.MakeRequests(&spec.BidRequest, reqInfo)
+
+	// Compare MakeRequests actual output versus expected values found in JSON file
+	assertErrorList(t, fmt.Sprintf("%s: MakeRequests", filename), errs, spec.MakeRequestErrors)
+	assertMakeRequestsOutput(t, filename, actualReqs, spec.HttpCalls)
+
+	// Assert no data races occur using original bidRequest copies of references and values
+	assertNoDataRace(t, deepBidReqCopy, shallowBidReqCopy, filename)
+
+	return actualReqs
+}
+
+func getDataRaceTestCopies(original *openrtb2.BidRequest) (*openrtb2.BidRequest, *openrtb2.BidRequest) {
+
+	// Save original bidRequest values to assert no data races occur inside MakeRequests latter
+	deepReqCopy := openrtb2.BidRequest{}
+	copier.Copy(&deepReqCopy, original)
+
+	// Shallow copy PBS core provides to adapters
+	shallowReqCopy := *original
+
+	// Save original []Imp elements to assert no data races occur inside MakeRequests latter
+	deepReqCopy.Imp = make([]openrtb2.Imp, len(original.Imp))
+	shallowReqCopy.Imp = make([]openrtb2.Imp, len(original.Imp))
+	for i := 0; i < len(original.Imp); i++ {
+		deepImpCopy := openrtb2.Imp{}
+		copier.Copy(&deepImpCopy, original.Imp[i])
+		deepReqCopy.Imp = append(deepReqCopy.Imp, deepImpCopy)
```

Comment:
Sorry, I think you explained this before. Why do we need to do make deep copies of the impression objects? Does `copier.Copy(&deepReqCopy, original)` not already do that?

Reply:
Because the deep copy version of this library was throwing false positives when comparing the `Ext` fields at every level. Switched to github.com/mohae/deepcopy to avoid it. The for loop could not be removed though because we still have to create the shallow copies of the `[]Imp` array elements.